Service class improvements

mwperina opened the pull request:
- **Fix race in bluechi-node-property-peer-ip test**
- **Use Path instance to hold service name and installation directory**
- **Use Service class in bluechi services cmd line config tests**
- **Track changed files at create_file in machine.py**

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/830

mwperina committed:
Use Path instance to hold service name and installation directory

Use Path instance instead of string to hodl service name and
installation directory inside Service class.

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/830
Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Fix race in bluechi-node-property-peer-ip test

Fixes rase in bluechi-node-property-peer-ip test around
last_seen_timestamp being updated before being checked

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Use Service class in bluechi services cmd line options tests

Use Service class to modify bluechi-agent or bluechi-controller systemd
service files to use additional command line options within existing
tests.

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi commented on tests/bluechi_test/machine.py:
```diff
@@ -103,15 +103,24 @@ def systemctl_start_and_wait(self, service_name: str, sleep_after: float = 0.0)
-        LOGGER.debug(f"Installing systemd service '{service.name}' to container path '{target_dir}'"
-                     f"with content:\n{service}")
-        self.create_file(target_dir, service.name, str(service))
+    def install_systemd_service(self, service: Service, backup_existing: bool = False, stop_at_cleanup: bool = True):
```
`stop_at_cleanup` isn't used here - and I'd remove it.
I think all services installed this way need to be stopped and their failure counter reset - including the bluechi services (they are stopped in `shutdown_bluechi` shortly before `teardown` anyway). 
Is there a reason not to stop installed services?

mwperina replied:
Done

engelmi commented on tests/bluechi_test/machine.py:
```diff
@@ -103,15 +103,24 @@ def systemctl_start_and_wait(self, service_name: str, sleep_after: float = 0.0)
+
+        LOGGER.debug(f"Installing systemd service '{service.get_name()}' to container path '{service.get_directory()}'"
+                     f"with content:\n{service.to_string()}")
+        self.create_file(service.get_directory(), service.get_name(), service.to_string())
```
I think it would be better to decide if a backup of the file should be created or not in `install_systemd_service`. So I'd remove the `backup_existing` parameter:

```suggestion
        LOGGER.debug(f"Installing systemd service '{service.get_name()}' to container path '{service.get_directory()}'"
                     f"with content:\n{service.to_string()}")
        self.create_file(service.get_directory(), service.get_name(), service.to_string())
```

and enhance the `create_file` function to check if there is already a file for the target path and, if yes, create and track a backup, e.g. like this:
```python
try:
    _, output = self.client.exec_run(f"[ -f {target_file} ] && echo 'exists'")
    if output == "exists":
        self._track_changed_file(target_dir, file_name)
    self.client.create_file(target_dir, file_name, content)
```

mwperina replied:
Done

engelmi commented on tests/bluechi_test/machine.py:
```diff
@@ -103,15 +103,24 @@ def systemctl_start_and_wait(self, service_name: str, sleep_after: float = 0.0)
+        self.systemctl.tracked_services.append(service.get_name())
+
+    def load_systemd_service(self, service: Service):
+        result, output = self.exec_run(f"cat {str(service.get_path())}")
```
Shouldn't this use `service.get_path().absolute()` or `service.get_path().resolve()`?

mwperina replied:
No need

engelmi commented on tests/bluechi_test/machine.py:
```diff
@@ -103,15 +103,24 @@ def systemctl_start_and_wait(self, service_name: str, sleep_after: float = 0.0)
-        self.systemctl.tracked_services.append(service.name)
+        self.systemctl.tracked_services.append(service.get_name())
+
+    def load_systemd_service(self, service: Service):
```
```suggestion
    def get_systemd_service_from_file(self, service_path: pathlib.Path) -> Service:
```
Passing the Service and updating it feels like C-style. What do you think about passing a `str` or `pathlib.Path` and return a new `Service` instance?

engelmi replied:
In the usage in `restart_with_config_file` it would then be:
```python
service = self.get_systemd_service_from_file(service_path)
```

mwperina replied:
Done

engelmi commented on tests/bluechi_test/machine.py:
```diff
@@ -142,16 +151,17 @@ def copy_machine_lib(self):
+                           service.get_option(Section.Service, Option.ExecStart) + f" -c {cfg_file_location}")
+
+        self.install_systemd_service(service, backup_existing=True, stop_at_cleanup=False)
+        self.systemctl.restart_unit(svc_name)
```
This function only makes sense for `bluechi-agent` and `bluechi-controller`. So I think it is implemented too general. What do you think about having:
```python
class BluechiService(Service):
# ...
class BluechiAgentService(BluechiService):
# ...
class BluechiControllerService(BluechiService):
#...

# and changing function to:
def restart_with_config_file(self, cfg_file_location: str, bluechi_service: BluechiService) -> None:
```

mwperina replied:
I don't see a point in creating subclasses, which don't add any logic, just predefined service location and name.

engelmi replied:
The point in this case is to limit the usage of `restart_with_config_file` to `bluechi-agent` and `bluechi-controller` - it makes no sense (and rather introduces bugs) to use any other service here and add `-c <path>` to its ExecStart. The `restart_with_config_file` function could also be part of the `BluechiService` class (and the machine will be passed as parameter) so it wouldn't be empty - and thinking about the other tests for the parameters (`-p` etc.) it would make sense to provide a function for each one on the bluechi class and encapsulate common functionality (like install + restart).

mwperina replied:
OK, I will try to take a look how to make restart_with_config_file usable for all usage (pretty much focus on updating the service file and restart a service, while service file changes would be done within each test)

engelmi replied:
I really like that the `restart_with_config_file` is now a simple load + install systemd unit

engelmi commented on tests/bluechi_test/service.py:
```diff
@@ -38,37 +46,61 @@ class Option(enum.Enum):
-        self.cp.optionxform = lambda option: option
-        self.name = name
+    """A base service class, which by default contains only predefined sections"""
+    def __init__(self, name: str = SERVICE_NAME, directory: str = SERVICE_DIRECTORY) -> None:
```
```suggestion
    def __init__(self, directory: str = SERVICE_DIRECTORY, name: str = SERVICE_NAME) -> None:
```
NIT: The reversed order feels more natural to me - but thats a personal preference.

mwperina replied:
Done

engelmi commented on tests/bluechi_test/service.py:
```diff
@@ -38,37 +46,61 @@ class Option(enum.Enum):
+        if not directory:
+            raise Exception(f"Service directory is '{directory}', but cannot be empty!")
+
+        self._svc_path = pathlib.Path(directory) / pathlib.Path(name)
```
```suggestion
        self._svc_path = pathlib.Path(directory).joinPath(pathlib.Path(name))
```
I don't know about any different behavior between those two approaches, though.

mwperina replied:
There shouldn't be any significant difference: https://github.com/python/cpython/blob/main/Lib/pathlib/_abc.py#L395-L407

engelmi commented on tests/bluechi_test/service.py:
```diff
@@ -38,37 +46,61 @@ class Option(enum.Enum):
     """A simple service, which is using /bin/true command"""
-    def __init__(self, name: str = "simple.service") -> None:
+    def __init__(self, name: str = "simple.service", directory: str = SERVICE_DIRECTORY) -> None:
         super(SimpleService, self).__init__(name)
```
```suggestion
        super(SimpleService, self).__init__(name, directory)
```
The directory parameter needs to be passed in as well.

mwperina replied:
Done

engelmi commented on tests/bluechi_test/service.py:
```diff
@@ -81,15 +113,15 @@ def __init__(self, name: str = "simple.service") -> None:
     """A simple service, which adds RemainAfterExit enabled to SimpleService"""
-    def __init__(self, name: str = "simple.service") -> None:
+    def __init__(self, name: str = "simple.service", directory: str = SERVICE_DIRECTORY) -> None:
         super(SimpleRemainingService, self).__init__(name)
```
```suggestion
        super(SimpleRemainingService, self).__init__(name, directory)
```
The directory parameter needs to be passed in as well.

mwperina replied:
Done

engelmi commented on tests/bluechi_test/service.py:
```diff
@@ -81,15 +113,15 @@ def __init__(self, name: str = "simple.service") -> None:
     """A simple service, which is using /bin/sleep command"""
-    def __init__(self, name: str = "sleeping.service") -> None:
+    def __init__(self, name: str = "sleeping.service", directory: str = SERVICE_DIRECTORY) -> None:
         super(SleepingService, self).__init__(name)
```
```suggestion
        super(SleepingService, self).__init__(name, directory)
```
The directory parameter needs to be passed in as well.

mwperina replied:
Done

mwperina committed:
Track changed files at create_file in machine.py

create_file is the only method we use to create new files on nodes, so
it's easiest to detect if created file already exists and if it's track
this file as being change by a test so the change can be reverted at the
test cleanup phase.

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi commented on tests/bluechi_test/machine.py:
```diff
@@ -41,6 +41,10 @@ def __init__(self, name: str, client: Client) -> None:
+            if output == "exists":
+                self._track_changed_file(target_dir, file_name)
+
             self.client.create_file(target_dir, file_name, content)
```
Looking at it now I realized that we would also need to move line for keeping track of created files into the try-catch-block and check if the file didn't exist before. If it existed before, we keep track of it as a changed file, otherwise its a new one (which will simply be deleted).
```python
self.client.create_file(target_dir, file_name, content)

if output != "exists":
    self.created_files.append(os.path.join(target_dir, file_name))

mwperina replied:
Great catch!

mwperina replied:
Fixed

engelmi reviewed (approved):
LGTM
